server/lomas_server/dp_queries/dp_libraries/opendp.py

```diff
@@ -81,28 +83,128 @@ def get_lf_domain(metadata):
+
+    # Global params
+    params = {}
     if "rows" in metadata:
+        params["max_partition_length"] = metadata["rows"]
+    if "max_ids" in metadata:
+        params["max_num_partitions"] = metadata["max_ids"]
+    if len(params) > 0:
         lf_domain = dp.domains.with_margin(
             lf_domain,
             by=[],
             public_info="lengths",
-            max_partition_length=metadata["rows"],
-            max_num_partitions=1,
+            **params,
         )
 
-    for name, series_info in metadata["columns"].items():
-        # TODO add max_partition_contributions, max_influenced_partitions
-        if "max_num_partitions" in series_info:
-            lf_domain = dp.domains.with_margin(
-                lf_domain,
-                by=[name],
-                public_info="lengths",
-                max_num_partitions=series_info["max_num_partitions"],
-            )
+    # Grouping logic (margin adaptation)
+    if by_config:
+        params = _update_params_by_grouping(metadata, by_config, params)
+        lf_domain = dp.domains.with_margin(
+            lf_domain,
+            by=by_config,
+            public_info="lengths",
+            **params,
+        )
 
     return lf_domain
 
 
+def _update_params_by_grouping(metadata, by_config, params):
+    """
+    Updates the parameters for margin adaptation based on
+    grouping configuration.
+
+    Args:
+        metadata (dict): The metadata dictionary.
+        by_config (list): Configuration for grouping.
+        params (dict): Current parameters dictionary to update.
+
+    Returns:
+        dict: Updated parameters dictionary.
+    """
+    if len(by_config) == 1:
+        series_info = metadata["columns"].get(by_config[0])
+        _single_group_update_params(metadata, series_info, params)
+    else:
+        _multiple_group_update_params(metadata, by_config, params)
+
+    params["max_partition_length"] = int(params["max_partition_length"])
+    return params
+
+
+def _single_group_update_params(metadata, series_info, params):
+    """
+    Updates parameters for single-column grouping configuration.
+
+    Args:
+        metadata (dict): The metadata dictionary.
+        series_info (dict): Metadata for the series (column).
+        params (dict): Current parameters dictionary to update.
+    """
+    params["max_partition_length"] = int(
+        metadata["rows"] * series_info.get("max_partition_length", 1)
+    )
+    params["max_num_partitions"] = series_info.get("cardinality")
+
+    if "max_influenced_partitions" in series_info:
+        params["max_influenced_partitions"] = min(
+            metadata["max_ids"],
+            series_info.get("max_influenced_partitions"),
+        )
+
+    if "max_partition_contributions" in series_info:
+        params["max_partition_contributions"] = min(
+            metadata["max_ids"],
+            series_info.get("max_partition_contributions"),
+        )
```

Comment:
I think it would be nice to have tests for these two option, like test on one file that has this metadata and one that doesn't. But I guess it can also be done after it is merged in the feature branch :)

Reply:
i totally agree